src/Mono.Android/Xamarin.Android.Net/AndroidMessageHandler.cs

```diff
@@ -0,0 +1,1018 @@
+					// When you use the parameter Java.Net.Proxy.NoProxy the system proxy is overriden. Leave the parameter out to respect the default settings.
+					java_connection = javaProxy == Java.Net.Proxy.NoProxy ? java_url.OpenConnection () : java_url.OpenConnection (javaProxy);
+				} else {
+					// In this case the consumer of this class has explicitly chosen to not use a proxy, so bypass the default proxy. The default value of UseProxy is true.
+					java_connection = java_url.OpenConnection (Java.Net.Proxy.NoProxy);
+				}
+
+				var httpsConnection = java_connection as HttpsURLConnection;
+				if (httpsConnection != null) {
+					IHostnameVerifier? hnv = GetSSLHostnameVerifier (httpsConnection);
+					if (hnv != null)
+						httpsConnection.HostnameVerifier = hnv;
+				}
+
+				if (ConnectTimeout != TimeSpan.Zero)
+					java_connection!.ConnectTimeout = checked ((int)ConnectTimeout.TotalMilliseconds);
+
+				if (ReadTimeout != TimeSpan.Zero)
+					java_connection!.ReadTimeout = checked ((int)ReadTimeout.TotalMilliseconds);
+
+				try {
+					HttpURLConnection httpConnection = await SetupRequestInternal (request, java_connection!).ConfigureAwait (continueOnCapturedContext: false);
+					HttpResponseMessage? response = await ProcessRequest (request, java_url, httpConnection, cancellationToken, redirectState).ConfigureAwait (continueOnCapturedContext: false);
+					if (response != null)
+						return response;
+
+					if (redirectState.NewUrl == null)
+						throw new InvalidOperationException ("Request redirected but no new URI specified");
+					request.Method = redirectState.Method;
+				} catch (Java.Net.SocketTimeoutException ex) when (JNIEnv.ShouldWrapJavaException (ex)) {
+					throw new WebException (ex.Message, ex, WebExceptionStatus.Timeout, null);
+				} catch (Java.Net.UnknownServiceException ex) when (JNIEnv.ShouldWrapJavaException (ex)) {
+					throw new WebException (ex.Message, ex, WebExceptionStatus.ProtocolError, null);
+				} catch (Java.Lang.SecurityException ex) when (JNIEnv.ShouldWrapJavaException (ex)) {
+					throw new WebException (ex.Message, ex, WebExceptionStatus.SecureChannelFailure, null);
+				} catch (Java.IO.IOException ex) when (JNIEnv.ShouldWrapJavaException (ex)) {
+					throw new WebException (ex.Message, ex, WebExceptionStatus.UnknownError, null);
+				}
+			}
+		}
+
+		internal virtual async Task <Java.Net.Proxy?> GetJavaProxy (Uri destination, CancellationToken cancellationToken)
+		{
+			var proxy = Java.Net.Proxy.NoProxy;
+
+			if (destination == null || Proxy == null) {
+				goto done;
+			}
+
+			Uri puri = Proxy.GetProxy (destination);
+			if (puri == null) {
+				goto done;
+			}
+
+			proxy = await Task <Java.Net.Proxy>.Run (() => {
+				// Let the Java code resolve the address, if necessary
+				var addr = new Java.Net.InetSocketAddress (puri.Host, puri.Port);
+				return new Java.Net.Proxy (Java.Net.Proxy.Type.Http, addr);
+			}, cancellationToken);
+
+		  done:
```

Comment:
Tab space label?!

Reply:
The original has:
https://github.com/xamarin/xamarin-android/blob/57636c8199efe724d1875f05d6b68051dfbcec02/src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs#L324-L327

I never use `goto` or labels, what is best here? Make it line up with the `return` statement?

Reply:
I think I should just remove all the `goto` here and put `return proxy` instead.